### ISSUE TITLE
[UI Testing] False positive when using `.isLoaded()` to verify a modal has been dismissed

### ISSUE THREAD

**Issue by thehenrybyrd** [woocommerce/woocommerce-ios#7526]
**Describe the bug**
It seems `.isLoaded()` isn’t verifying what we expected, particularly when working with modals. If a modal is opened on top of a screen, `.isLoaded()` is true for the underlying screen. We expected that `.isLoaded()` would only be true for the top-most layer, i.e. the modal.

Because `.isLoaded()` is still true for a screen under a modal, we cannot verify that the modal is dismissed. 
In our case, we wanted to verify that tapping “done” in the Order Editing flow dismisses the modal, and the Single Order Screen is once again visible. Instead, our UI tests pass even when the modal is not successfully dismissed.

**To Reproduce**
To reproduce, stop a modal from being dismissed, and note that the screen below is still “loaded”, though it is not visible.
Here’s a specific example:
In `test_edit_existing_order`, we use `.closeEditingFlow()` and then `.verifySingleOrderScreenLoaded()`, to check that the single order screen loads after we complete the editing flow. Here’s how to break the “done” button on that modal, in order to reproduce this issue:

1. Break the “done” button by commenting out dismissHandler() on line 172 of OrderForm, [as shown here](https://href.li/?https://github.com/woocommerce/woocommerce-ios/compare/trunk%E2%80%A6update/edit-order-ui-test#diff-2e1179be83959416087ddd5227efbd0be1f15ba739eadfd62943a0c548d8a54dR172).
2. Run `test_edit_existing_order`.
3. Watch the simulator, and note that the “done” button is tapped, but modal is not dismissed.
4. Note that `.verifySingleOrderScreenLoaded()` passes.

**Temporary solution**: For the moment, we’ve gotten around this in our branch by adding `XCTAssertFalse(doneButton.exists)` to our `closeEditingFlow()` function – to check that the modal isn’t still there.

However, this is a larger issue with any test that relies on .isLoaded() to check for a screen after a modal is dismissed. If we use the XCTAssertFalse workaround, we will always have to remember this when testing with modals. So, we want to suggest updating .isLoaded(), part of the ScreenObject dependency.

Ideally, we should find a way to makes sure .isLoaded() only checks the visible layer. One possibility we thought of is adding .isHittable to the predicate within waitForScreen (at [this line](https://href.li/?https://github.com/Automattic/ScreenObject/blob/e27405a65672c62e5a055697eafdeaf073ea63ff/Sources/ScreenObject/ScreenObject.swift#L100)). One drawback of isHittable is that it could cause flakiness depending on screen size or scroll status – but this may not be a big issue depending on the tests effected.

Internal discussion at pdcxQM-193-p2.

**Comment by jostnes** [comment]
Some updates after early investigation, 3 things need to happen before we can close this:

- [x] Update screenObject to use the above recommendation (use `isHittable`), update package release version
- [x] Update screenObject package in WCiOS repo to use the updated version and run all UI tests to ensure all still work

This is tested, minimal code changes needed to make it work.
- [x]  Update screenObject package in WPiOS repo to use the updated version and run all UI tests to ensure all still work 

This is tested, there are two affected test cases right now after the change, `testAddSelfHostedSiteAfterWPcomLogin` and `testTabBarNavigation`. However, `testAddSelfHostedSiteAfterWPcomLogin` seems to always fail when tested locally with/without change so I'll create a separate issue for that. `testTabBarNavigation` looks to be looking for a `table` element, I suspect that `.isHittable` doesn't work for that element but further testing is needed to confirm. 

There are other tests that are failing locally but pass after a retry in WPiOS, will have to run more tests to ensure that this change is not the cause of the flakiness.

**Comment by jostnes** [comment]
@thehenrybyrd this issue should be resolved now, the 3 PRs for the fix are linked above. Thank you for raising your concern on this and if you see anything weird after this please continue to share with the team.
